arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+
+    runWork(work);
+
+    stealAttempts = 0;
+    maxStealAttempts = totalMaxStealAttempts;
+  };
+
+  while (!stop.load(std::memory_order_relaxed)) {
+    std::unique_lock guard(mutex);
+    if (!_queue.empty()) {
+      auto item = std::move(_queue.front());
+      TRI_ASSERT(item != nullptr);
+      _queue.pop_front();
+      queueSize.store(_queue.size(), std::memory_order_relaxed);
+      guard.unlock();
+
+      doWork(*item);
+    } else if (pushQueue.load(std::memory_order_relaxed) != nullptr) {
+      auto item = pushQueue.exchange(nullptr, std::memory_order_acquire);
+      if (item != nullptr) {
+        auto work = pushMany(item);
+        guard.unlock();
+        doWork(*work);
+      }
+    } else if (stealAttempts > maxStealAttempts) {
+      // nothing to work -> go to sleep
+      goToSleep(guard);
+      if (maxStealAttempts > 2) {
+        maxStealAttempts /= 2;
+      }
+      stealAttempts = 0;
+    } else {
+      guard.unlock();
+      if (stealAttempts == 0) {
+        // try to insert ourselves in the hint chain
+        // threads with lower IDs spin longer, so we prefer to have them at the
+        // front of the chain, provided they are not sleeping already.
+        // We simply limit ourselves to 4 steps before giving up (note that the
+        // chain can contain loops)
+        auto* hint = &pool.hint;
+        for (unsigned i = 0; i < 4; ++i) {
+          auto h = hint->load(std::memory_order_relaxed);
+          auto& ts = pool._threadStates[h];
+          if (h > id || ts->sleeping.load(std::memory_order_relaxed)) {
+            prevHint.store(h, std::memory_order_relaxed);
+            hint->store(id, std::memory_order_relaxed);
+            break;
+          }
+          hint = &ts->prevHint;
+        }
+      }
+      auto work = stealWork();
+      if (work) {
+        doWork(*work);
+      } else {
+        basics::cpu_relax();
+        ++stealAttempts;
+      }
+    }
+  }
+}
+
+void WorkStealingThreadPool::ThreadState::goToSleep(
+    std::unique_lock<std::mutex>& guard) noexcept {
+  sleeping.store(true, std::memory_order_seq_cst);
+  // the store and load both need to be seq_cst to ensure proper ordering with
+  // the push that wakes us up
+  if (pushQueue.load(std::memory_order_seq_cst) == nullptr) {
+    pool.lastSleepIdx.store(id, std::memory_order_relaxed);
+    pool.numSleeping.fetch_add(1, std::memory_order_relaxed);
+    cv.wait_for(guard, std::chrono::milliseconds{100});
+    pool.numSleeping.fetch_sub(1, std::memory_order_relaxed);
+  }
+  sleeping.store(false, std::memory_order_relaxed);
+}
+
+auto WorkStealingThreadPool::ThreadState::pushMany(WorkItem* item)
+    -> std::unique_ptr<WorkItem> {
+  TRI_ASSERT(item != nullptr);
+  while (item->next != nullptr) {
+    auto next = item->next;
+    item->next = nullptr;
+    _queue.push_front(std::unique_ptr<WorkItem>(item));
+    item = next;
+    TRI_ASSERT(item != nullptr);
+  }
+  queueSize.store(_queue.size(), std::memory_order_relaxed);
```

Comment:
nit: we could save this store operation in case we did not push anything into `_queue` in this method. this would save an atomic store, but not sure if it's worth optimizing it.

Reply:
`pushMany` should not be called with a nullptr (note the assertion), so we should always add at least one item to the queue.

Reply:
I agree. But even if `pushMany` is called with a non-nullptr `item`, it is possible and valid that `item->next` is a nullptr, and then the queue size will not be modified. Thus we also don't need to perform the atomic store on the queue size variable. But this is an optimization and not necessary for correctness.

Reply:
Ah, right, because we keep the last item to be work on immediately. However, I don't think this is worth the optimization.

Reply:
Fine for me too.